File: python/acp-protocol/src/acp_protocol/subjects.py
"""NATS subject rendering from the shared vocabulary (schemas/subjects.json).

Mirrors the TypeScript binding exactly; the cross-language contract tests
assert both render identical strings for identical inputs.
"""

import json
import re
from functools import cache
from importlib import resources
from typing import Any

_TOKEN_RE = re.compile(r"^[a-z0-9_.-]+$")
_EVENT_TYPE_RE = re.compile(r"^[a-z_]+(\.[a-z_]+)+$")


@cache
def _data() -> dict[str, Any]:
    ref = resources.files("acp_protocol").joinpath("schemas/subjects.json")
    loaded: dict[str, Any] = json.loads(ref.read_text(encoding="utf-8"))
    return loaded


def _token(name: str, value: str) -> str:
    # `.`, `*`, `>` are structural in NATS; a caller-supplied ID containing
    # them could widen a subscription, so reject rather than sanitize.
    if not _TOKEN_RE.match(value) or ".." in value:
        raise ValueError(f"invalid subject token {name}={value!r}: must match {_TOKEN_RE.pattern}")
    return value


def _verb(entity: str, verb: str) -> str:
    verbs: list[str] = _data()["entities"][entity]["verbs"]
    if verb not in verbs:
        raise ValueError(
            f"unknown {entity} verb {verb!r}: closed vocabulary is [{', '.join(verbs)}]"
        )
    return verb


def task(tenant: str, task_id: str, verb: str) -> str:
    return f"acp.{_token('tenant', tenant)}.task.{_token('task_id', task_id)}.{_verb('task', verb)}"


def agent(tenant: str, agent_id: str, verb: str) -> str:
    return (
        f"acp.{_token('tenant', tenant)}.agent."
        f"{_token('agent_id', agent_id)}.{_verb('agent', verb)}"
    )


def audit(tenant: str, event_type: str) -> str:
    if not _EVENT_TYPE_RE.match(event_type):
        raise ValueError(f"invalid audit event_type {event_type!r}")
    return f"acp.{_token('tenant', tenant)}.audit.{event_type}"


def audit_corpus(tenant: str, source_id: str) -> str:
    return f"acp.{_token('tenant', tenant)}.audit.corpus.{_token('source_id', source_id)}"


def ingest(tenant: str, source_id: str) -> str:
    return f"acp.{_token('tenant', tenant)}.ingest.{_token('source_id', source_id)}"


def telemetry(tenant: str, signal: str) -> str:
    return f"acp.{_token('tenant', tenant)}.telemetry.{_token('signal', signal)}"


def registry(agent_id: str, verb: str) -> str:
    return f"acp.platform.registry.{_token('agent_id', agent_id)}.{_verb('registry', verb)}"


def control(verb: str) -> str:
    return f"acp.platform.control.{_verb('control', verb)}"

File: packages/protocol/src/subjects.ts
import { subjectsData } from './generated/subjects-data.js';

const TOKEN_RE = /^[a-z0-9_.-]+$/;

/**
 * Validates a single subject token. NATS treats `.`, `*`, and `>` as
 * structural; letting caller-supplied IDs smuggle them in would let a tenant
 * widen its own subscription — so this throws rather than sanitizes.
 * (`event_type` values like `task.submitted` are multi-token by design and
 * are validated against the closed vocabulary instead.)
 */
function checkToken(name: string, value: string): string {
  if (!TOKEN_RE.test(value) || value.includes('..')) {
    throw new Error(
      `invalid subject token ${name}=${JSON.stringify(value)}: must match ${TOKEN_RE.source}`,
    );
  }
  return value;
}

function checkVerb(entity: keyof typeof subjectsData.entities, verb: string): string {
  const verbs: readonly string[] = subjectsData.entities[entity].verbs;
  if (!verbs.includes(verb)) {
    throw new Error(
      `unknown ${entity} verb ${JSON.stringify(verb)}: closed vocabulary is [${verbs.join(', ')}]`,
    );
  }
  return verb;
}

export type TaskVerb = (typeof subjectsData.entities.task.verbs)[number];
export type AgentVerb = (typeof subjectsData.entities.agent.verbs)[number];
export type RegistryVerb = (typeof subjectsData.entities.registry.verbs)[number];
export type ControlVerb = (typeof subjectsData.entities.control.verbs)[number];
export type SvcName = (typeof subjectsData.entities.svc.services)[number];

export const subjects = {
  task(tenant: string, taskId: string, verb: TaskVerb): string {
    return `acp.${checkToken('tenant', tenant)}.task.${checkToken('task_id', taskId)}.${checkVerb('task', verb)}`;
  },
  agent(tenant: string, agentId: string, verb: AgentVerb): string {
    return `acp.${checkToken('tenant', tenant)}.agent.${checkToken('agent_id', agentId)}.${checkVerb('agent', verb)}`;
  },
  audit(tenant: string, eventType: string): string {
    // event_type is validated against the closed enum in the audit-event
    // schema; the subject uses it verbatim (dots become subject tokens).
    if (!/^[a-z_]+(\.[a-z_]+)+$/.test(eventType)) {
      throw new Error(`invalid audit event_type ${JSON.stringify(eventType)}`);
    }
    return `acp.${checkToken('tenant', tenant)}.audit.${eventType}`;
  },
  auditCorpus(tenant: string, sourceId: string): string {
    return `acp.${checkToken('tenant', tenant)}.audit.corpus.${checkToken('source_id', sourceId)}`;
  },
  ingest(tenant: string, sourceId: string): string {
    return `acp.${checkToken('tenant', tenant)}.ingest.${checkToken('source_id', sourceId)}`;
  },
  telemetry(tenant: string, signal: string): string {
    return `acp.${checkToken('tenant', tenant)}.telemetry.${checkToken('signal', signal)}`;
  },
  registry(agentId: string, verb: RegistryVerb): string {
    return `acp.platform.registry.${checkToken('agent_id', agentId)}.${checkVerb('registry', verb)}`;
  },
  control(verb: ControlVerb): string {
    return `acp.platform.control.${checkVerb('control', verb)}`;
  },
  svc(service: SvcName, method: string): string {
    const services: readonly string[] = subjectsData.entities.svc.services;
    if (!services.includes(service)) {
      throw new Error(
        `unknown platform service ${JSON.stringify(service)}: closed vocabulary is [${services.join(', ')}]`,
      );
    }
    return `acp.platform.svc.${service}.${checkToken('method', method)}`;
  },
} as const;
